File: zkpschnorrproofs/src/commonMain/kotlin/io/github/suwasto/zkpschnorrproofs/ZKPUtils.kt
package io.github.suwasto.zkpschnorrproofs

import com.ionspin.kotlin.bignum.integer.BigInteger

object ZKPUtils {

    internal fun modularExponentiation(base: BigInteger, exponent: BigInteger, mod: BigInteger): BigInteger {
        var result = BigInteger.ONE
        var newBase = base % mod
        var newExponent = exponent

        while (newExponent > BigInteger.ZERO) {
            if (newExponent % BigInteger.TWO == BigInteger.ONE) {
                result = (result * newBase) % mod
            }
            newExponent /= BigInteger.TWO
            newBase = (newBase * newBase) % mod
        }
        return result
    }

}

File: kmpzkpjs/src/ZKPUtils.ts
import bigInt, { BigInteger } from "big-integer";

export class ZKPUtils {
  static modularExponentiation(base: BigInteger, exponent: BigInteger, mod: BigInteger): BigInteger {
    let result = bigInt.one;
    let newBase = base.mod(mod);
    let newExponent = exponent;

    while (newExponent.greater(0)) {
      if (newExponent.mod(2).equals(1)) {
        result = result.multiply(newBase).mod(mod);
      }
      newExponent = newExponent.divide(2);
      newBase = newBase.multiply(newBase).mod(mod);
    }

    return result;
  }
}
